### PR TITLE
[Regex] Capture groups typehint

### PR DESCRIPTION
**Pull request by veewee**
This PR adds better return types for Regex\capture_groups introduced in:
https://github.com/azjezz/psl/pull/151

### CODE REVIEW

File: src/Plugin.php [hunk 1]
```diff
@@ -17,5 +17,7 @@ public function __invoke(RegistrationInterface $registration, ?SimpleXMLElement
 
         $registration->registerHooksFromClass(EventHandler\Type\Optional\FunctionReturnTypeProvider::class);
         $registration->registerHooksFromClass(EventHandler\Type\Shape\FunctionReturnTypeProvider::class);
+
+        // TODO : Awaiting #3's getHooks() function
```
Comment (veewee) [inline src/Plugin.php]: Awaiting #3

Comment (azjezz) [reply src/Plugin.php]: #3 is merged, you can rebase now :)

File: src/EventHandler/Regex/CaptureGroups/FunctionReturnTypeProvider.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        $call_args = $event->getCallArgs();
+
+        $argument = $call_args[0] ?? null;
+        if (null === $argument) {
+            return self::fallbackType();
+        }
+
+        $argument_value = $argument->value;
+        $type = $statements_source->getNodeTypeProvider()->getType($argument_value);
+        if (null === $type) {
```
Comment (azjezz) [inline src/EventHandler/Regex/CaptureGroups/FunctionReturnTypeProvider.php:35]:
```suggestion
        $type = Argument::getType($event->getCallArgs(), $event->getStatmentsSource(), 0);
        if (null === $type) {
```

File: src/EventHandler/Regex/CaptureGroups/FunctionReturnTypeProvider.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        ];
+        foreach ($capture_groups->properties as $value) {
+            $type = array_values($value->getAtomicTypes())[0] ?? null;
+            if (!$type instanceof Type\Atomic\TLiteralInt && !$type instanceof Type\Atomic\TLiteralString) {
+                return self::fallbackType();
+            }
+
+            $name = $type->value;
+
+            $properties[$name] = $string();
```
Comment (azjezz) [inline src/EventHandler/Regex/CaptureGroups/FunctionReturnTypeProvider.php:57]:
```suggestion
            $properties[$name] = Type::getString();
```

File: src/EventHandler/Regex/CaptureGroups/FunctionReturnTypeProvider.php [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+        $string = static fn (): Type\Union => new Type\Union([new Type\Atomic\TString()]);
+        $properties = [
+            0 => $string()
+        ];
+        foreach ($capture_groups->properties as $value) {
+            $type = array_values($value->getAtomicTypes())[0] ?? null;
+            if (!$type instanceof Type\Atomic\TLiteralInt && !$type instanceof Type\Atomic\TLiteralString) {
+                return self::fallbackType();
+            }
```
Comment (azjezz) [inline src/EventHandler/Regex/CaptureGroups/FunctionReturnTypeProvider.php:53]: do you think we should add an invalid argument issue here if the argument values are not literal?

Comment (veewee) [reply src/EventHandler/Regex/CaptureGroups/FunctionReturnTypeProvider.php:53]: The argument is typed as `list<array-key>`
So it is not really a valid case I suppose?

Comment (azjezz) [reply src/EventHandler/Regex/CaptureGroups/FunctionReturnTypeProvider.php:53]: we can still trigger errors from here, but not sure it's worth it.

**Pull request merged**